apps/tidalgraph/tidalgraph.star

```diff
@@ -0,0 +1,340 @@
+                default = options[0].value,
+                options = options,
+            ),
+            schema.Text(
+                id = "stationid",
+                name = "Station ID",
+                desc = "NOAA Station ID",
+                icon = "water",
+            ),
+        ],
+    )
+
+def get_tide_data(stationId, date, interval):
+    url = "https://api.tidesandcurrents.noaa.gov/api/prod/datagetter?begin_date=%s&end_date=%s&station=%s&product=predictions&datum=MLLW&time_zone=lst_ldt&interval=%s&units=english&application=DataAPI_Sample&format=json" % (date, date, stationId, interval)
+
+    data = cache.get("%s-%s" % (stationId, date))
+    if data != None:
+        data = json.decode(data)
+        print("Hit! Displaying cached data.")
+    else:
+        print("Miss! Calling tide API: %s" % url)
+        response = http.get(url)
+        if response.status_code != 200 or "error" in response.json():
+            print("tide request failed with status %d" % response.status_code)
+            return None
+        data = response.json()
+        cache.set("%s-%s" % (stationId, date), response.body(), ttl_seconds = 86400)
+
+    return data
+
+def get_date(current_offset_time):
+    return current_offset_time.format("20060102")
+
+def get_data_points(today):
+    data_points = []
+    max = 0
+    min = 0
+
+    for p in today["predictions"]:
+        time = p["t"][11:].split(":")
+        hours = int(time[0]) + (int(time[1]) / 60)
+        value = float(p["v"])
+        data_points.append((hours, value))
+        if value > max:
+            max = value
+        if value < min:
+            min = value
+
+    return {"points": data_points, "max_val": max, "min_val": min}
+
+def get_current_state(data, current_time):
+    current = None
+    next = None
+    for p in data["predictions"]:
+        if p["t"][11:] < current_time.format("15:04"):
+            current = p["v"]
+        elif next == None:
+            next = p["v"]
+
+    if current > next:
+        return "%s Receding" % current
+    else:
+        return "%s Rising" % current
+
+def calculate_hours(timestamp):
+    time = timestamp.split(":")
+    return int(time[0]) + (int(time[1]) / 60)
+
+def get_station_timezone(id):
+    url = "https://api.tidesandcurrents.noaa.gov/mdapi/prod/webapi/stations/%s.json" % id
+    zone = cache.get(id)
+    if zone != None:
+        print("Hit! Displaying cached data.")
+        return zone
+    else:
+        print("Miss! Calling station API: %s" % url)
+        response = http.get(url)
+        if response.status_code != 200:
+            print("station request failed with status %d" % response.status_code)
+            return None
+        zone = response.json()["stations"][0]["timezone"]
+        if zone in TIMEZONE_MAP:
+            tz = TIMEZONE_MAP[zone]
+            print(tz)
+            cache.set(id, str(tz), ttl_seconds = 86400)
+            return tz
+        else:
+            return "unsupported_timezone:%s" % zone
```

Comment:
Not sure how often this happens, but could be good to cache something here as well to keep the app from unnecessarily trying to re-request the data.

Reply:
yes, good call! This _shouldn't_ happen very often, but I could happen. I added cacheing to the unsupported timezones

Reply:
tyvm!